src/computability/regular_expressions.lean

```diff
@@ -0,0 +1,295 @@
+computer science such as the POSIX standard.
+-/
+
+universe u
+
+variables {α : Type u} [dec : decidable_eq α]
+
+/-- This is the definition of regular expressions. The names used here is to mirror the definition
+  of a Kleene algebra (https://en.wikipedia.org/wiki/Kleene_algebra).
+  `RZero` matches nothing
+  `RNull` matches only the empty string
+  `RChar a` matches only the string 'a'
+  `RStar M` matches any finite concatenation of strings which match `M`
+  `RPlus M N` matches anything which match `M` or `N`
+  `RComp M N` matches `x ++ y` if `x` matches `M` and `y` matches `N` -/
+inductive regular_expression (α : Type u) : Type (u+1)
+| RZero : regular_expression
+| RNull : regular_expression
+| RChar : α → regular_expression
+| RStar : regular_expression → regular_expression
+| RPlus : regular_expression → regular_expression → regular_expression
+| RComp : regular_expression → regular_expression → regular_expression
+
+namespace regular_expression
+
+instance regular_expression_inhabited : inhabited (regular_expression α) := ⟨ RZero ⟩
+
+instance : has_add (regular_expression α) := ⟨RPlus⟩
+instance : has_mul (regular_expression α) := ⟨RComp⟩
+instance : has_one (regular_expression α) := ⟨RNull⟩
+instance : has_zero (regular_expression α) := ⟨RZero⟩
+
+/-- `match_null M` is true if and only if `M` matches the empty string -/
+def match_null : regular_expression α → bool
+| RZero := ff
+| RNull := tt
+| (RChar _) := ff
+| (RStar M) := tt
+| (RPlus M N) := M.match_null || N.match_null
+| (RComp M N) := M.match_null && N.match_null
+
+include dec
+
+/-- `M.feed a` matches `x` if `M` matches `"a" ++ x` -/
+def feed : regular_expression α → α → regular_expression α
+| RZero _ := RZero
+| RNull _ := RZero
+| (RChar a₁) a₂ := if a₁ = a₂ then RNull else RZero
+| (RStar M) a := RComp (feed M a) (RStar M)
+| (RPlus M N) a := RPlus (feed M a) (feed N a)
+| (RComp M N) a :=
+  ite M.match_null (RPlus (RComp (feed M a) N) (feed N a)) (RComp (feed M a) N)
+
+/-- `M.rmatch x` is true if and only if `M` matches `x` -/
+def rmatch : regular_expression α → list α → bool
+| M [] := match_null M
+| M (a::as) := rmatch (M.feed a) as
+
+/-- Two regular expressions are equivalent if they match exactly the same strings -/
+def equiv (P Q : regular_expression α) : Prop := ∀ x, P.rmatch x ↔ Q.rmatch x
+
+local infix ` ≈ ` := equiv
+
+@[refl] lemma equiv_refl (P : regular_expression α) : P ≈ P := λ x, by refl
+@[symm] lemma equiv_symm (P Q : regular_expression α) : P ≈ Q → Q ≈ P := λ h x, (h x).symm
+@[trans] lemma equiv_trans (P Q R : regular_expression α) : P ≈ Q → Q ≈ R → P ≈ R :=
+  λ h₁ h₂ x, iff.trans (h₁ x) (h₂ x)
+
+@[simp] lemma equiv_def (P Q : regular_expression α) : P ≈ Q ↔ ∀ x, P.rmatch x ↔ Q.rmatch x :=
+  by refl
+
+lemma RZero_rmatch (x : list α) : rmatch RZero x = ff :=
+begin
+  induction x,
+    rw [rmatch, match_null],
+  rwa [rmatch, feed],
+end
+
+@[simp] lemma zero_rmatch (x : list α) : rmatch 0 x = ff := RZero_rmatch x
```

Comment:
Try to stick to a single notation.  Either `0` or `zero`, and then add a simp lemma `zero = 0` that simplifies to the preferred version.